#### Data Analysis with pandas

**Use-case Scenarion:**

You are provided with data (test_data.xlsx) from a retail shop that consists of 8 columns(Invoice, StockCode, Description, Quantity, InvoiceDate, Price, Customer ID, and Country).
You are supposed to analyze data in indexes between (433200, 437200)! Do not forget to clean the data of NaNs, as failing to do so may render your answers invalid!

*Fill in your answers to the text below, as if it were your report from the data.*



1.   The amount of **unique invoices** in the provided dataset for indexes (433200, 437200) is equal to --?-- (integer).

2.   The **country with the most orders** is --?-- (string) with the **amount** of --?-- (integer).

3.   The **country with the least orders** is --?-- (string) with the amount of --?-- (integer).
4.   The **highest total price per item** in one order was --?--  (float) in order with **id** --?-- (integer).
5.   The **second highest spending customer** with id --?-- The second highest spending customer with id --?--(integer) **times**.


**Hints:**
Total price per item is calculated as price times quantity.
Methods to consider: *dropna, group_by, sort_values*

In [124]:
# Loading Lib
import pandas as pd
import numpy as np

In [125]:
# Loading Data Note: Important to spacify the sheet you are working on
data1 = pd.read_excel('/content/test_data.xlsx', sheet_name = 'Year 2009-2010')

In [126]:
# slicing our desired Data
data_df = data1.iloc[433200: 437200]
data_df

Invoice StockCode                          Description  Quantity  \
433200  530811    16206B            RED PURSE WITH PINK HEART         3   
433201  530811     22662          LUNCH BAG DOLLY GIRL DESIGN         1   
433202  530811     20727              LUNCH BAG  BLACK SKULL.         2   
433203  530811     20726                   LUNCH BAG WOODLAND         3   
433204  530811     22382           LUNCH BAG SPACEBOY DESIGN          2   
...        ...       ...                                  ...       ...   
437195  531210    85071B   RED CHARLIE+LOLA PERSONAL DOORSIGN         1   
437196  531210    85071A  BLUE CHARLIE+LOLA PERSONAL DOORSIGN         1   
437197  531210     22898        CHILDRENS APRON APPLES DESIGN         1   
437198  531210     22899         CHILDREN'S APRON DOLLY GIRL          1   
437199  531210     21788                   KIDS RAIN MAC BLUE         2   

               InvoiceDate  Price  Customer ID         Country  
433200 2010-11-04 13:04:00   1.25      13137.0  United Kingdom  
433201 2010-11-04 13:04:00   1.65      13137.0  United Kingdom  
433202 2010-11-04 13:04:00   1.65      13137.0  United Kingdom  
433203 2010-11-04 13:04:00   1.65      13137.0  United Kingdom  
433204 2010-11-04 13:04:00   1.65      13137.0  United Kingdom  
...                    ...    ...          ...             ...  
437195 2010-11-05 16:07:00   2.95      13263.0  United Kingdom  
437196 2010-11-05 16:07:00   2.95      13263.0  United Kingdom  
437197 2010-11-05 16:07:00   1.95      13263.0  United Kingdom  
437198 2010-11-05 16:07:00   2.10      13263.0  United Kingdom  
437199 2010-11-05 16:07:00   0.85      13263.0  United Kingdom  

[4000 rows x 8 columns]

In [127]:
# we must have information about the data that we are working with
data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 433200 to 437199
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Invoice      4000 non-null   object        
 1   StockCode    4000 non-null   object        
 2   Description  3917 non-null   object        
 3   Quantity     4000 non-null   int64         
 4   InvoiceDate  4000 non-null   datetime64[ns]
 5   Price        4000 non-null   float64       
 6   Customer ID  3028 non-null   float64       
 7   Country      4000 non-null   object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 250.1+ KB


In [128]:
data_df.describe()

Quantity        Price   Customer ID
count  4000.000000  4000.000000   3028.000000
mean      5.012500     3.741615  15154.674373
std     177.398748     8.183365   1865.822694
min   -8800.000000     0.000000  12393.000000
25%       1.000000     1.250000  13615.000000
50%       3.000000     2.510000  15023.000000
75%      10.000000     4.250000  16919.000000
max    1152.000000   341.280000  18259.000000

In [129]:
# Droping Null Values in axis=0 means droping whole Row which is containg it
df = data_df.dropna(axis=0)
df

Invoice StockCode                          Description  Quantity  \
433200  530811    16206B            RED PURSE WITH PINK HEART         3   
433201  530811     22662          LUNCH BAG DOLLY GIRL DESIGN         1   
433202  530811     20727              LUNCH BAG  BLACK SKULL.         2   
433203  530811     20726                   LUNCH BAG WOODLAND         3   
433204  530811     22382           LUNCH BAG SPACEBOY DESIGN          2   
...        ...       ...                                  ...       ...   
437195  531210    85071B   RED CHARLIE+LOLA PERSONAL DOORSIGN         1   
437196  531210    85071A  BLUE CHARLIE+LOLA PERSONAL DOORSIGN         1   
437197  531210     22898        CHILDRENS APRON APPLES DESIGN         1   
437198  531210     22899         CHILDREN'S APRON DOLLY GIRL          1   
437199  531210     21788                   KIDS RAIN MAC BLUE         2   

               InvoiceDate  Price  Customer ID         Country  
433200 2010-11-04 13:04:00   1.25      13137.0  United Kingdom  
433201 2010-11-04 13:04:00   1.65      13137.0  United Kingdom  
433202 2010-11-04 13:04:00   1.65      13137.0  United Kingdom  
433203 2010-11-04 13:04:00   1.65      13137.0  United Kingdom  
433204 2010-11-04 13:04:00   1.65      13137.0  United Kingdom  
...                    ...    ...          ...             ...  
437195 2010-11-05 16:07:00   2.95      13263.0  United Kingdom  
437196 2010-11-05 16:07:00   2.95      13263.0  United Kingdom  
437197 2010-11-05 16:07:00   1.95      13263.0  United Kingdom  
437198 2010-11-05 16:07:00   2.10      13263.0  United Kingdom  
437199 2010-11-05 16:07:00   0.85      13263.0  United Kingdom  

[3028 rows x 8 columns]

In [130]:
# Question 1.
len(df.Invoice.unique())

171

In [131]:
# max(df.Country)

In [132]:
# min(df.Country)

In [133]:
# created new col with Total sale of one item
df['Total'] = df.Quantity * df.Price

In [134]:
# Question 2
max_cn = df[(df.Country == max(df.Country))]['Total'].count()
print(f"Country {max(df.Country)} with maximum orders of : {max_cn}")

Country United Kingdom with maximum orders of : 2688


In [135]:
# Question 3
min_cn = df[(df.Country == min(df.Country))]['Total'].count()
print(f"Country {min(df.Country)} with minimum orders of : {min_cn}")

Country Australia with minimum orders of : 5


In [136]:
# single item with highest Total price
# df[df.Total == df['Total'].max()]
# Highest ordered Quantity
# df[df.Quantity == df['Quantity'].max()]

In [137]:
# Question 4.
largest_invoice = df.groupby(df.Invoice)['Total'].sum().sort_values(ascending=False)
largest_invoice

Invoice
531204     5321.16
530970     2570.77
531069     2496.96
530966     2292.18
531138     2147.82
            ...   
C531001     -67.50
C530910     -70.80
C531133     -95.64
C530971    -102.00
C530832    -168.36
Name: Total, Length: 171, dtype: float64

In [138]:
# Question 5
# find Id of the
orders_sum = df.groupby(df['Customer ID'])['Total'].sum().sort_values(ascending=False).nlargest(3)
print("\nids with  orders amount: ", orders_sum)

sec_ind = orders_sum.index[1]
print(sec_ind)

norders = df[df['Customer ID'] == sec_ind].Invoice.nunique()
print("\nNo of the Times the seconed Highest spending customer ordered = ", norders)

ids with  orders amount
 Customer ID
17389.0    6412.14
12477.0    2570.77
17949.0    2496.96
Name: Total, dtype: float64
12477.0

No of the Times the seconed Highest spending customer ordered =  1


#Report:

1.   The amount of **unique invoices** in the provided dataset for indexes (433200, 437200) is equal to 171 (integer).

2.   The **country with the most orders** is ***United Kingdom*** (string) with the **amount** of 2688 (integer).

3.   The **country with the least orders** is ***Australia*** (string) with the **amount** of 5 (integer).
4.   The **highest total price per item** in one order was 5321.16  (float) in order with **id** 531204 (integer).
5.   The **second highest spending customer** with id 12477 The second highest spending customer with id 1 (integer) **times**.

In [9]:
import warnings
warnings.filterwarnings('ignore')
import pandas as pd

data = pd.read_excel('/content/test_data.xlsx', sheet_name = 'Year 2009-2010')
data = data.iloc[433200 : 437200]
selected_data  = data.dropna(axis=0)


In [10]:

# Answering the questions:

# 1. The amount of unique invoices
unique_invoices = selected_data['Invoice'].nunique()
selected_data['Total'] = selected_data.Quantity * selected_data.Price
# 2. The country with the most orders
most_orders_country = selected_data['Country'].value_counts().idxmax()
most_orders_count = selected_data['Country'].value_counts().max()

# 3. The country with the least orders
least_orders_country = selected_data['Country'].value_counts().idxmin()
least_orders_count = selected_data['Country'].value_counts().min()

# 4. The highest total price per item in one order
max_price_per_item_order = selected_data.groupby('Invoice')['Total'].sum().max()
max_price_per_item_order_id = selected_data.groupby('Invoice')['Total'].sum().idxmax()

# 5. The second highest spending customer
top_customers = selected_data.groupby('Customer ID')['Total'].sum().sort_values(ascending=False)

# 5. The second highest spending customer and the total number of times they ordered
second_highest_spending_customer = top_customers.index[1]
second_highest_spending_customer_orders = selected_data[selected_data['Customer ID'] == second_highest_spending_customer]['Invoice'].nunique()

# Displaying the updated result

print(f"1. The amount of unique invoices is {unique_invoices}.")
print(f"2. The country with the most orders is {most_orders_country} with {most_orders_count} orders.")
print(f"3. The country with the least orders is {least_orders_country} with {least_orders_count} orders.")
print(f"4. The highest total price per item in one order was {max_price_per_item_order:.2f} in order with id {max_price_per_item_order_id}.")
print(f"5. The second highest spending customer has id {second_highest_spending_customer} and ordered {second_highest_spending_customer_orders} times.")

1. The amount of unique invoices is 171.
2. The country with the most orders is United Kingdom with 2688 orders.
3. The country with the least orders is Spain with 1 orders.
4. The highest total price per item in one order was 5321.16 in order with id 531204.
5. The second highest spending customer has id 12477.0 and ordered 1 times.
